Pod VM image handling fix

bpradipt opened the pull request:
Fixes: #[KATA-2958](https://issues.redhat.com//browse/KATA-2958) #[KATA-2962](https://issues.redhat.com//browse/KATA-2962)

snir911 commented on controllers/image_generator.go:
```diff
@@ -640,6 +647,26 @@ func (r *ImageGenerator) createImageConfigMapFromFile() error {
+			igLogger.Info("Setting IMAGE_GALLERY_NAME", "image_gallery_name", image_gallery_name)
+		}
+
+		// Add tags to IMAGE_GALLERY_TAGS
```
what is the use case of having multiple tags?

bpradipt replied:
For OSC, we only need one tag (ie, clusterId). But then realised that end users might want to have specific tags like with all other resources for billing etc. And since I'm adding support for tags, felt it's better to have this flexibility

snir911 commented on controllers/openshift_controller.go:
```diff
@@ -90,7 +90,7 @@ const (
 // +kubebuilder:rbac:groups=apps,resources=daemonsets/finalizers,resourceNames=manager-role,verbs=update
 // +kubebuilder:rbac:groups=node.k8s.io,resources=runtimeclasses,verbs=get;list;watch;create;update;patch;delete
-// +kubebuilder:rbac:groups=config.openshift.io,resources=clusterversions,verbs=get
+// +kubebuilder:rbac:groups=config.openshift.io,resources=clusterversions,verbs=get;list;watch
```
note that it doesn't seems like manifests were regenerated in this PR (maybe these rules are already included)

bpradipt replied:
Yes, the manifests file are now ignored. Take a look at .gitignore

snir911 commented on controllers/utils.go:
```diff
@@ -202,3 +202,15 @@ func getCloudProviderFromInfra(c client.Client) (string, error) {
+
+	// Return first 8 characters of the cluster id
+	return string(clusterVersion.Spec.ClusterID[:8]), nil
+}
```
Why not doing this directly in the scripts IIRC we do have kubectl/oc there available there

bpradipt replied:
You are right, this can be done in the scripts as well. The reason for doing it here is to feed the input config for gallery (azure-podvm-image-cm) via the operator. The azure-podvm-image-cm is input only config and I consciously avoided updating it in the script.

snir911 commented on config/peerpods/podvm/azure-podvm-image-handler.sh:
```diff
@@ -410,7 +410,64 @@ function add_image_id_annotation_to_peer_pods_cm() {
+    fi
+
+    # Add IMAGE_GALLERY_NAME annotation to peer-pods-cm configmap
+    kubectl annotate configmap peer-pods-cm -n openshift-sandboxed-containers-operator \
```
can you explain the rational set the annotations, I think i may have missed something

bpradipt replied:
The peer-pods-cm is the single source of truth for dynamically updated info. Hence it's used to store the image as well as the gallery name. The podvm-handling.md explains it as well.

snir911 commented on controllers/image_generator.go:
```diff
@@ -640,6 +647,26 @@ func (r *ImageGenerator) createImageConfigMapFromFile() error {
 		return err
 	}
 
+	// For Azure provider set the IMAGE_GALLERY_NAME if its empty
```
AFAIU it will reach this point only when the default CM is used (and created from file) and not when the use is providing the a CM which misses the IMAGE_GALLERY_NAME or set it to ""

bpradipt replied:
right..

snir911 replied:
So if the image CM exist already (637) shouldn't we make sure gallery image name is set? IIUC it will fail on gallery creation? is it ok to assume if user create he's own CM he's know he should set it

bpradipt replied:
If user creates the CM, then user should set the gallery name. We'll update in the README or instructions which provides steps on manual creation of the image creation job (outside of the operator)

snir911 commented on config/peerpods/podvm/podvm-builder.sh:
```diff
@@ -180,12 +206,27 @@ function delete_podvm_image_gallery() {
+    echo "IMAGE_GALLERY_NAME is not set in peer-pods-cm. Skipping image gallery deletion"
+    exit 0
+  fi
+
```
Am i getting this right? Image gallery will get deleted only if was set by OSC, based on the annotation we set?
And the image itself will get deleted based on IMAGE_ID value (if not changed) right? however, we do have the image_id as annotation as well, would it make sense to use the same mechanism?

bpradipt replied:
The way to check if the image gallery was created by OSC is the tag. The annotation just mentions the image gallery being used. The value in the annotation could be a pre-existing gallery that is entered by the admin user in the azure-podvm-image-cm

snir911 replied:
Sorry, i don't get why the difference in implementation is needed, ie. why galleries are deleted based on the annotations values and images are deleted based on the CM key value (data)?

bpradipt replied:
The reason being a gallery can hold multiple images which is not the case for an image. But you do have a point. I think having a tag for the image is also required to ensure we only delete the image created by OSC and not any external image

bpradipt replied:
Oh forgot to mention that the image is created by packer and the only way to add tag is after creation of the image. But I think as you suggested I can use annotation as the source of truth for OSC created resources and use the same logic for gallery and image.

bpradipt replied:
I have changed it to use the gallery value from annotation and avoid using tags completely.

snir911 replied:
> Oh forgot to mention that the image is created by packer and the only way to add tag is after creation of the image. But I think as you suggested I can use annotation as the source of truth for OSC created resources and use the same logic for gallery and image. 

oh i missed that, now it make sense, thanks

bpradipt committed:
Delete Image gallery when deleting the Image

PodVM image deletion function has been extended to also support
deletion of gallery via a CLI option. This is used by OSC to ensure
deletion of the OSC created gallery during podvm image deletion. This
ensures if gallery is created as part of kataconfig creation, then
the same gallery is deleted as well during kataconfig deletion

Fixes: #KATA-2958

Signed-off-by: Pradipta Banerjee <pradipta.banerjee@gmail.com>

gkurz commented on config/peerpods/podvm/azure-podvm-image-handler.sh:
```diff
@@ -410,7 +410,64 @@ function add_image_id_annotation_to_peer_pods_cm() {
         error_exit "Failed to add the image id as annotation to peer-pods-cm configmap"
 
-    echo "Image id added as annotation to peer-pods-cm configmap successfully"
+    echo "Image id annotation added as annotation to peer-pods-cm configmap successfully"
```
This line doesn't seem to require change since it already mentions the annotation or am I missing something ?

bpradipt replied:
Yeah..

bpradipt committed:
Create and Delete a unique gallery name if required

If IMAGE_GALLERY_NAME in azure-podvm-image-cm is empty, then a
gallery name will be created with the pattern
"PodVMGallery_$clusterid. Note that only the first 8 chars of the
clusterid is used.
IMAGE_GALLERY_NAME annotation is added to peer-pods-cm configMap
if gallery is created by OSC job

FIXES: #KATA-2958

Signed-off-by: Pradipta Banerjee <pradipta.banerjee@gmail.com>

bpradipt committed:
Add a README describing the podvm handling

Signed-off-by: Pradipta Banerjee <pradipta.banerjee@gmail.com>

bpradipt committed:
Fix delete_image_using_id method

az cli doesn't delete image with version ids using the
az image delete command. Instead we need to use
az sig image-version delete command for sig images.

Ref:
https://learn.microsoft.com/en-us/cli/azure/sig/image-version?view=azure-cli-latest#az-sig-image-version-delete

Fixes: #KATA-2962

Signed-off-by: Pradipta Banerjee <pradipta.banerjee@gmail.com>

bpradipt committed:
Remove delete_image_using_name function

This is not used anywhere

Signed-off-by: Pradipta Banerjee <pradipta.banerjee@gmail.com>

snir911 reviewed (approved):
LGTM 
Thanks!

bpradipt committed:
Update README for podvm image creation

Signed-off-by: Pradipta Banerjee <pradipta.banerjee@gmail.com>

gkurz commented on config/peerpods/podvm/podvm-builder.sh:
```diff
@@ -87,9 +86,28 @@ function create_podvm_image() {
 
+  local max_args=3
+  if (("${#@}" > max_args)); then
+    error_exit "Too many arguments passed to delete_podvm_image (max ${max_args} allowed)"
```
Why checking the number of arguments ?

bpradipt replied:
This is to ensure the function is used in the intended way with the required number of arguments. Ideally I wanted to have a similar check for all the functions to ensure only the accepted number of arguments are passed but that's best suited for another PR.
If this doesn't make sense I'm happy to remove it.

bpradipt replied:
@gkurz can we move ahead with this or you want this check to be removed ?

gkurz replied:
FWIW usual scheme is to check that the function is passed arguments it understands rather than the number of arguments.

gkurz commented on config/peerpods/podvm/azure-podvm-image-handler.sh:
```diff
@@ -47,7 +47,10 @@ function verify_vars() {
     [[ -z "${IMAGE_DEFINITION_PUBLISHER}" ]] && error_exit "IMAGE_DEFINITION_PUBLISHER is empty"
     [[ -z "${IMAGE_DEFINITION_OFFER}" ]] && error_exit "IMAGE_DEFINITION_OFFER is empty"
+
+    [[ -z "${IMAGE_GALLERY_NAME_PREFIX}" ]] && error_exit "IMAGE_GALLERY_NAME_PREFIX is empty"
```
What's the motivation behind this check ? The variable isn't used in the script AFAICT (like most of the variables verified in this function).

bpradipt replied:
Afaicr I added this to ensure when the script is executed via the operator (as part of job execution) it's always set

bpradipt replied:
I went through this again and some of the older iterations of this code. I don't see a logical reason to keep this check anymore. Instead this variable is not needed and I can set the prefix directly in the controller code.

gkurz commented on config/peerpods/podvm/azure-podvm-image-handler.sh:
```diff
@@ -413,6 +412,63 @@ function add_image_id_annotation_to_peer_pods_cm() {
+        "LATEST_IMAGE_ID-" ||
+        error_exit "Failed to delete the image id annotation from peer-pods-cm configmap"
+
+    echo "Image id added as annotation to peer-pods-cm configmap successfully"
```
```suggestion
    echo "Image id annotation deleted from peer-pods-cm configmap successfully"
```

gkurz commented on config/peerpods/podvm/azure-podvm-image-handler.sh:
```diff
@@ -555,6 +612,7 @@ function delete_image_gallery() {
     # Check if the gallery has any image versions
     get_all_image_ids
 
+    # This will set the IMAGE_ID_LIST variable
```
`IMAGE_ID_LIST` is set by `get_all_image_ids`. Maybe move the comment before the call to the function.

gkurz commented on controllers/image_generator.go:
```diff
@@ -644,6 +651,17 @@ func (r *ImageGenerator) createImageConfigMapFromFile() error {
+	if r.provider == AzureProvider {
+		if cm.Data["IMAGE_GALLERY_NAME"] == "" {
+			image_gallery_name := cm.Data["IMAGE_GALLERY_NAME_PREFIX"] + "_" + r.clusterId
+			cm.Data["IMAGE_GALLERY_NAME"] = image_gallery_name
```
Should we check that the generated name is valid ?

```
  # Note that the gallery name must be unique across the subscription and not exceed 80 characters.
  # Also, the allowed characters are English alphanumeric characters,
  # with underscores and periods allowed in the middle
```

bpradipt replied:
Based on this comment https://github.com/openshift/sandboxed-containers-operator/pull/404#discussion_r1618405423
if I change the code to 
```
const osc_gallery_name_prefix = PodVMGallery
image_gallery_name :=  osc_gallery_name_prefix + "_" + r.clusterId
```
This will always be valid as clusterId is only the first 8 chars. So imho no need to recheck if the generated name is valid. Wdyt ?

gkurz commented on config/peerpods/podvm/podvm-builder.sh:
```diff
@@ -180,12 +206,27 @@ function delete_podvm_image_gallery() {
   else
     /scripts/azure-podvm-image-handler.sh -G
   fi
+
```
nit: empty line

gkurz commented on config/peerpods/podvm/podvm-builder.sh:
```diff
@@ -87,9 +86,28 @@ function create_podvm_image() {
 # These are checked in individual cloud provider scripts and if not set, the script will exit
+# Accepts two optional arguments
+# -f : force delete the image
+# -g : delete the image gallery
```
Please add `[-g]` to the usage at https://github.com/openshift/sandboxed-containers-operator/pull/404/commits/8caf648d06893619d282ae13b4e085cbe97f82a1#diff-6544cd9ec466fe5e11123ce3f835bf520a39dd0e245eb676ccff7bebda9c4757R233.

gkurz commented on config/peerpods/podvm/azure-podvm-image-handler.sh:
```diff
@@ -666,7 +666,7 @@ function delete_image_using_id() {
 
     # Delete the image
-    az image delete --ids "${IMAGE_ID}" ||
+    az image-version delete --ids "${IMAGE_ID}" ||
```
Providing some reference URLs to clarify this change would be appreciated (but not strongly required).

bpradipt replied:
Looking at your comment made me realise a typo. The command should be `az sig image-version delete ..`. I have fixed it.

gkurz commented on config/peerpods/podvm/podvm-builder.sh:
```diff
@@ -189,7 +224,7 @@ function delete_podvm_image_gallery() {
 
 function display_usage() {
-  echo "Usage: $0 {create|delete [-f]|delete-gallery [-f]}"
+  echo "Usage: $0 {create|delete [-f]|delete-gallery [-g]}"
```
```suggestion
  echo "Usage: $0 {create|delete [-f] [-g]|delete-gallery [-f]}"
```

bpradipt replied:
Thanks.. Not sure what I was doing when making this change :-(

gkurz reviewed (approved):
/lgtm @bpradipt ! Just one last nit to fix in the usage string and we're good to go.
